src/Storages/StorageTimeSeries.cpp

```diff
@@ -0,0 +1,524 @@
+#include <Storages/StorageTimeSeries.h>
+
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypeDateTime64.h>
+#include <DataTypes/DataTypeMap.h>
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/DataTypeTuple.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Interpreters/Context.h>
+#include <Interpreters/DatabaseCatalog.h>
+#include <Interpreters/InterpreterCreateQuery.h>
+#include <Interpreters/InterpreterDropQuery.h>
+#include <Parsers/ASTDropQuery.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTIdentifier.h>
+#include <Parsers/ASTTargetTables.h>
+#include <Storages/AlterCommands.h>
+#include <Storages/StorageFactory.h>
+#include <base/insertAtEnd.h>
+#include <filesystem>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int INCORRECT_QUERY;
+}
+
+
+namespace
+{
+    namespace fs = std::filesystem;
+
+    const ColumnsDescription & getColumnsDescription()
+    {
+        static const ColumnsDescription columns_description{
+            { "id",                 std::make_shared<DataTypeUInt128>()                                                                         },
+            { "metric_name",        std::make_shared<DataTypeString>()                                                                          },
+            { "tags",               std::make_shared<DataTypeMap>(std::make_shared<DataTypeString>(), std::make_shared<DataTypeString>())       },
+            { "time_series",        std::make_shared<DataTypeArray>(std::make_shared<DataTypeTuple>(
+                                        DataTypes{std::make_shared<DataTypeDateTime64>(3), std::make_shared<DataTypeFloat64>()},
+                                        Strings{"timestamp", "value"})) },
+        };
+        return columns_description;
+    };
+
+    const ColumnsDescription & getColumnsDescription(TargetTableKind inner_table_kind)
+    {
+        switch (inner_table_kind)
+        {
+            case TargetTableKind::kData:
+            {
+                static const ColumnsDescription columns_description{
+                    { "id",        std::make_shared<DataTypeUInt128>()     },
+                    { "timestamp", std::make_shared<DataTypeDateTime64>(3) },
+                    { "value",     std::make_shared<DataTypeFloat64>()     },
+                };
+                return columns_description;
+            }
+
+            case TargetTableKind::kTags:
+            {
+                static const ColumnsDescription columns_description{
+                    { "id",          std::make_shared<DataTypeUInt128>()                                                                   },
+                    { "metric_name", std::make_shared<DataTypeString>()                                                                    },
+                    { "tags",        std::make_shared<DataTypeMap>(std::make_shared<DataTypeString>(), std::make_shared<DataTypeString>()) },
```

Comment:
If it expected that metrics can be filtered by tags often why not have it in EAV format?

```
id UInt128,
metric_name String,
tag_name String
tag_value String

ORDER BY (metric_name, tag_name)
```

Reply:
id is a hash of metric name combined with a sorted list of {tag_name, tag_value} pairs

Reply:
```id UInt128,
metric_name LowCardinality(String),
tag_name LowCardinality(String) ~
tag_value String
```
Why not `tag_value LowCartdinality(String)` then?

Reply:
I understand from where it's coming.

But 2 things concerns me:

1. Heavily unbalanced columns: id ~16 bytes, metric_name ~ 10-20 bytes, tags potentially hundreds bytes
2. How filtering will work by tags.

Reply:
I thought about the structure of the `tags` table more. For supporting the Prometheus read protocol it's actually more convenient to have the structure of the `tags` table in the manner where all the tags related to a specific `ID` are described in one row. So
```
id UInt128,
metric_name LowCardinality(String),
tag_name LowCardinality(String) ~
tag_value String
```
is not really suitable.

The following structure
```
id UInt128,
metric_name String,
tags Map(String, String)
```
must be ok, and also I've added support for putting some tags into separate columns of the `tags` table. For example, the following statement
```
CREATE TABLE ts ENGINE=TimeSeries() SETTINGS tags_to_columns={'job': 'job', 'instance':'instance'}
```
will create the `tags` table with columns
```
id UInt128,
metric_name String,
job String,
instance String,
tags Map(String, String)
```

Reply:
https://github.com/ClickHouse/ClickHouse/pull/47045

What if make buckets for map column?

As bucketed map type wasn't implemented yet

Reply:
> https://github.com/ClickHouse/ClickHouse/pull/47045
As bucketed map type wasn't implemented yet

Bucketed maps will be supported of course.

> What if make buckets for map column? (ie have X map columns, and distribute tags across them using hash function

That's exactly what a bucketed map does. It's better to wait until https://github.com/ClickHouse/ClickHouse/pull/47045 will be ready.

Reply:
> It's better to wait until https://github.com/ClickHouse/ClickHouse/pull/47045 will be ready.

It looks kinda death, but ok :) 

Anyway explicit materialization of certain columns is OK.

How query builder will work with them? 
It will be some specific implementation for Prometheus Protocol or general, like "finish constraint optimization in production" https://github.com/ClickHouse/ClickHouse/issues/33544 ?